File: protocol-autoresearch/experiments/tier2_007_framework_integrator_session_token_offchain_jwt/sdk-python/bolyra/session.py
"""Off-chain session token (JWT) after on-chain handshake verification.

Uses EdDSA (Ed25519) via PyJWT + cryptography. Mirrors the TypeScript
``mintSessionToken`` / ``verifySessionToken`` API in ``@bolyra/sdk``.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Optional

import jwt
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)


# ── Types ──────────────────────────────────────────────────────────────────


@dataclass(frozen=True)
class SessionClaims:
    """Decoded session token claims."""

    nullifier_hash: str
    scope_commitment: str
    session_nonce: str
    iat: int
    exp: int


@dataclass
class HandshakeVerifyResult:
    """Result of a verifyHandshake() call."""

    valid: bool
    nullifier_hash: str
    scope_commitment: str
    session_nonce: str


class BolyraSessionError(Exception):
    """Raised when session token minting or verification fails."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


# ── Constants ──────────────────────────────────────────────────────────────

DEFAULT_TTL_SECONDS = 300
MIN_TTL_SECONDS = 60
MAX_TTL_SECONDS = 900
DEFAULT_ISSUER = "bolyra.ai"


# ── Public API ─────────────────────────────────────────────────────────────


def mint_session_token(
    verify_result: HandshakeVerifyResult,
    signer_key: Ed25519PrivateKey,
    *,
    ttl_seconds: int = DEFAULT_TTL_SECONDS,
    issuer: str = DEFAULT_ISSUER,
) -> str:
    """Mint a session JWT after a successful verifyHandshake().

    Args:
        verify_result: The handshake verification output.
        signer_key: An Ed25519 private key for signing.
        ttl_seconds: Token lifetime in seconds (60–900). Default: 300.
        issuer: JWT issuer claim. Default: 'bolyra.ai'.

    Returns:
        A signed JWT string.

    Raises:
        BolyraSessionError: If the handshake was invalid or TTL is out of range.
    """
    if not verify_result.valid:
        raise BolyraSessionError(
            "INVALID_TOKEN",
            "Cannot mint session token from invalid handshake result",
        )

    if ttl_seconds < MIN_TTL_SECONDS or ttl_seconds > MAX_TTL_SECONDS:
        raise BolyraSessionError(
            "INVALID_TOKEN",
            f"TTL must be between {MIN_TTL_SECONDS}s and {MAX_TTL_SECONDS}s, got {ttl_seconds}s",
        )

    now = int(time.time())

    payload = {
        "nullifierHash": verify_result.nullifier_hash,
        "scopeCommitment": verify_result.scope_commitment,
        "sessionNonce": verify_result.session_nonce,
        "iat": now,
        "exp": now + ttl_seconds,
        "iss": issuer,
    }

    return jwt.encode(
        payload,
        signer_key,
        algorithm="EdDSA",
        headers={"typ": "JWT"},
    )


def verify_session_token(
    token: str,
    public_key: Ed25519PublicKey,
    *,
    expected_issuer: str = DEFAULT_ISSUER,
) -> SessionClaims:
    """Verify a session token off-chain.

    Args:
        token: The JWT string to verify.
        public_key: The Ed25519 public key corresponding to the signer.
        expected_issuer: Expected issuer claim. Default: 'bolyra.ai'.

    Returns:
        Decoded SessionClaims.

    Raises:
        BolyraSessionError: On invalid signature, expiry, or missing claims.
    """
    try:
        payload = jwt.decode(
            token,
            public_key,
            algorithms=["EdDSA"],
            issuer=expected_issuer,
            options={"require": ["exp", "iat", "iss"]},
        )
    except jwt.ExpiredSignatureError:
        raise BolyraSessionError("TOKEN_EXPIRED", "Session token has expired")
    except jwt.InvalidSignatureError:
        raise BolyraSessionError("INVALID_SIGNATURE", "JWT signature verification failed")
    except jwt.PyJWTError as exc:
        raise BolyraSessionError("INVALID_TOKEN", f"JWT verification failed: {exc}")

    nullifier_hash = payload.get("nullifierHash")
    scope_commitment = payload.get("scopeCommitment")
    session_nonce = payload.get("sessionNonce")

    if not all(
        isinstance(v, str) for v in (nullifier_hash, scope_commitment, session_nonce)
    ):
        raise BolyraSessionError(
            "CLAIMS_TAMPERED",
            "Required claims (nullifierHash, scopeCommitment, sessionNonce) missing or invalid",
        )

    return SessionClaims(
        nullifier_hash=nullifier_hash,  # type: ignore[arg-type]
        scope_commitment=scope_commitment,  # type: ignore[arg-type]
        session_nonce=session_nonce,  # type: ignore[arg-type]
        iat=payload["iat"],
        exp=payload["exp"],
    )

File: protocol-autoresearch/experiments/tier2_004_framework_integrator_session_token_offchain_jwt/sdk/src/session.ts
/**
 * Off-chain JWT session token from a verified Bolyra handshake.
 *
 * After a single on-chain verifyHandshake(), the verifier mints a short-lived
 * JWT (ES256 / P-256 ECDSA) binding the nullifierHash and scopeCommitment.
 * Subsequent calls present only the JWT — reducing per-call overhead from
 * ~200ms proof verification to a single ECDSA check (~0.5ms).
 *
 * @module session
 */

import { SignJWT, jwtVerify } from 'jose';

// ── Types ────────────────────────────────────────────────────────────────────

/** Subset of verifyHandshake() output needed for token minting. */
export interface VerifiedHandshakeProof {
  nullifierHash: string;
  scopeCommitment: string;
}

/** Options for handshakeToSessionToken(). */
export interface SessionTokenOptions {
  /** Token lifetime in seconds. Default: 3600 (1 hour). */
  ttlSeconds?: number;
  /** JWT audience claim. Default: '*'. */
  audience?: string;
}

/** Decoded session token claims. */
export interface SessionTokenClaims {
  nullifier: string;
  scope: string;
  expiry: number;
}

export type SessionErrorCode =
  | 'INVALID_SIGNATURE'
  | 'TOKEN_EXPIRED'
  | 'MISSING_CLAIMS'
  | 'INVALID_TOKEN';

export class BolyraSessionError extends Error {
  constructor(public readonly code: SessionErrorCode, message: string) {
    super(message);
    this.name = 'BolyraSessionError';
  }
}

// ── Constants ─────────────────────────────────────────────────────────────────

const DEFAULT_TTL_SECONDS = 3600;
const ISSUER = 'bolyra';
const DEFAULT_AUDIENCE = '*';

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * Mint a session JWT from a verified handshake proof.
 *
 * @param proof - Output from verifyHandshake() containing nullifierHash and scopeCommitment.
 * @param signingKey - An ES256 (P-256) private CryptoKey.
 * @param options - Optional TTL and audience overrides.
 * @returns Compact-serialized JWT string.
 */
export async function handshakeToSessionToken(
  proof: VerifiedHandshakeProof,
  signingKey: CryptoKey,
  options: SessionTokenOptions = {},
): Promise<string> {
  if (!proof.nullifierHash || !proof.scopeCommitment) {
    throw new BolyraSessionError(
      'INVALID_TOKEN',
      'Proof must contain nullifierHash and scopeCommitment',
    );
  }

  const ttl = options.ttlSeconds ?? DEFAULT_TTL_SECONDS;
  const audience = options.audience ?? DEFAULT_AUDIENCE;

  const jwt = await new SignJWT({
    sub: proof.nullifierHash,
    scope: proof.scopeCommitment,
  })
    .setProtectedHeader({ alg: 'ES256', typ: 'JWT' })
    .setIssuedAt()
    .setExpirationTime(`${ttl}s`)
    .setIssuer(ISSUER)
    .setAudience(audience)
    .sign(signingKey);

  return jwt;
}

/**
 * Verify and decode a Bolyra session JWT.
 *
 * @param jwt - The compact JWT string.
 * @param pubKey - The ES256 (P-256) public CryptoKey that corresponds to the signer.
 * @returns Decoded claims: nullifier, scope, and expiry.
 * @throws {BolyraSessionError} On invalid signature, expiry, or missing claims.
 */
export async function verifySessionToken(
  jwt: string,
  pubKey: CryptoKey,
): Promise<SessionTokenClaims> {
  let payload: Record<string, unknown>;
  try {
    const result = await jwtVerify(jwt, pubKey, {
      issuer: ISSUER,
      algorithms: ['ES256'],
    });
    payload = result.payload as Record<string, unknown>;
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : String(err);
    if (message.includes('expired') || message.includes('exp')) {
      throw new BolyraSessionError('TOKEN_EXPIRED', 'Session token has expired');
    }
    throw new BolyraSessionError('INVALID_SIGNATURE', `JWT verification failed: ${message}`);
  }

  const nullifier = payload.sub;
  const scope = payload.scope;
  const expiry = payload.exp;

  if (typeof nullifier !== 'string' || typeof scope !== 'string' || typeof expiry !== 'number') {
    throw new BolyraSessionError(
      'MISSING_CLAIMS',
      'Required claims (sub, scope, exp) missing or invalid',
    );
  }

  return { nullifier, scope, expiry };
}
